### ISSUE TITLE
Elasticsearch exception No mapping found for [date.keyword] in order to sort on

### ISSUE THREAD

**Issue by indiepopart** [jhipster/generator-jhipster#17785]
##### **Overview of the issue**

Search for entities that had paginate in the JDL, after sorting by date, fail with the following exception:

```
Suppressed: org.elasticsearch.ElasticsearchException: Elasticsearch exception [type=query_shard_exception, reason=No mapping found for [date.keyword] in order to sort on]
```

##### **Motivation for or Use Case**

In entities that had `paginate`in the JDL, after sorting the table by a non-text field, subsequent search fails, as the query builder assumes the sort field has multi-field mapping, and has a keyword field for sorting. Keyword field is only created automatically for text fields.

##### **Reproduce the error**

```
jhipster jdl blog-reactive-ms.jdl
cd gateway
./mvnw -ntp -Pprod verify jib:dockerBuild
cd ../blog
./mvnw -ntp -Pprod verify jib:dockerBuild
cd ../docker-compose
docker compose up
```

1. Go to localhost:8080, sign in
2. Go to entities -> Blog
3. Create a Blog
4. Go to entities -> Post
5. Create a Post
6. Sort the table by Date
7. Search for a Post -> Internal Server Error

##### **Related issues**

Related to #17018

##### **JHipster Version(s)**

```
INFO! Using JHipster version installed globally
7.6.0
```

##### **JHipster configuration**

  <details>
  <summary>JDL definitions</summary>
  <pre>
application {
  config {
    baseName gateway
    reactive true
    packageName com.okta.developer.gateway
    applicationType gateway
    authenticationType oauth2
    buildTool maven
    clientFramework vue
    prodDatabaseType postgresql
    serviceDiscoveryType eureka
    testFrameworks [cypress]
  }
  entities Blog, Post, Tag
}
application {
  config {
    baseName blog
    reactive true
    packageName com.okta.developer.blog
    applicationType microservice
    authenticationType oauth2
    buildTool maven
    databaseType mongodb
    devDatabaseType mongodb
    prodDatabaseType mongodb
    enableHibernateCache false
    serverPort 8081
    serviceDiscoveryType eureka
    searchEngine elasticsearch
  }
  entities Blog, Post, Tag
  service * with serviceClass
  search * with elasticsearch
}
entity Blog {
  name String required minlength(3)
  handle String required minlength(2)
}
entity Post {
  title String required
  content TextBlob required
  date Instant required
}
entity Tag {
  name String required minlength(2)
}
relationship ManyToOne {
  Blog{user(login)} to User
  Post{blog(name)} to Blog
}
relationship ManyToMany {
  Post{tag(name)} to Tag{post}
}
paginate Post, Tag with pagination
microservice Blog, Post, Tag with blog
deployment {
  deploymentType docker-compose
  appsFolders [gateway, blog]
  dockerRepositoryName "jgarbarino"
}

  </pre>
  </details>

##### **Browsers and Operating System**

Ubuntu 18.04.6 LTS

- [ x ] Checking this box is mandatory (this is just to show you read everything)

**Comment by mraible** [comment]
@indiepopart Have you figured out a workaround for this that we might want to implement?

**Comment by indiepopart** [comment]
@mraible I could not work on this yet, let me get back next week

**Comment by maznag** [comment]
Hi @indiepopart, @mraible  the search is fail for all sort field not only for date.  Also, we have this problem only for app generated with option "reactive true", that generate this part of code that use `SortToSortBuilderListConverter` to build sort fields :

    List<SortBuilder<?>> builders = new SortToSortBuilderListConverter().convert(pageable.getSort());
        NativeSearchQueryBuilder queryBuilder = new NativeSearchQueryBuilder().withQuery(queryStringQuery(query))
            .withPageable(PageRequest.of(pageable.getPageNumber(), pageable.getPageSize()));

        queryBuilder.withSorts(builders); 
        NativeSearchQuery nativeSearchQuery = queryBuilder.build();

In `SortToSortBuilderListConverter` we have:

     sort.stream().forEach(order -> {
            String property = order.getProperty() + ".keyword";
            SortOrder sortOrder = SortOrder.fromString(order.getDirection().name());
            builders.add(new FieldSortBuilder(property).order(sortOrder));
        });

With this implementation, the search request with ElasticSearch will fail if there is no mapping associated with a field. As describe here [ignoring_unmapped_fields](https://www.elastic.co/guide/en/elasticsearch/reference/7.15/sort-search-results.html#_ignoring_unmapped_fields), the unmapped_type option allows to ignore fields that have no mapping and not sort by them. The value of this parameter is used to determine what sort values to emit.

if we want to continue with this implementation we should add the unmappedType as follow. But we should generate Field type `XXXXX` for each field in sort query using `org.springframework.data.elasticsearch.annotations.FieldType`

      sort.stream().forEach(order -> {
            String property = order.getProperty() + ".keyword";
            SortOrder sortOrder = SortOrder.fromString(order.getDirection().name());
            builders.add(new FieldSortBuilder(property).order(sortOrder).unmappedType("XXXXX");
        });

To simplify, I propose to use `org.springframework.data.elasticsearch.core.query.NativeSearchQuery` and `org.elasticsearch.index.query.queryStringQuery` that works well without adding any more code to generate fields type and unmappedType :

NativeSearchQuery nativeSearchQuery = new NativeSearchQuery(queryStringQuery(query));
        nativeSearchQuery.setPageable(pageable);


Here is a simple PR #18155 to fix this issue.

@indiepopart can you test again with this PR ?

<details>
<summary>To test this I generated simply the monolith with this jdl model</summary>

<pre>
entity Blog {
  name String required minlength(3)
  handle String required minlength(2)
}
entity Post {
  title String required
  content TextBlob required
  date Instant required
}
entity Tag {
  name String required minlength(2)
}
relationship ManyToOne {
  Blog{user(login)} to User
  Post{blog(name)} to Blog
}
relationship ManyToMany {
  Post{tag(name)} to Tag{post}
}

service Blog, Post, Tag with serviceClass
search Blog, Post, Tag with elasticsearch
paginate Post, Tag with pagination

</pre>
</details>

**Comment by indiepopart** [comment]
I think the fix restores the original implementation, #17018 might be happening again for text fields.
